cgo binaries have shared object dependency on libstdc++

If I build a cgo binary using `go build`, it only has a shared object dependency on the basic C libraries:
```console
$ make WHAT=cmd/kubelet
...
$ ldd _output/local/bin/linux/amd64/kubelet
        linux-vdso.so.1 (0x00007ffc6ed10000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f3c8737a000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f3c87176000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3c86dd7000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3c87597000)
```

If I build this same using bazel + rules_go (0.12.1, but I'm guessing newer versions are similar), there is an added dependency on `libstdc++` (and `libgcc_s`, `libm`):
```
$ bazel build cmd/kubelet
...
$ ldd bazel-bin/cmd/kubelet/linux_amd64_stripped/kubelet
        linux-vdso.so.1 (0x00007ffe67bb1000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fd02faf1000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd02f7ed000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fd02f5d0000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fd02f3cc000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd02f02d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fd02fe76000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fd02ee15000)
```

This is problematic if we try to include this binary in a docker container, say `busyboxy:glibc`, which has `libc.so` but not `libstdc++.so`.

@bentheelder ran the bazel build with debugging information (`bazel build -s //cmd/kubelet`), which eventually showed the final linking step:
```
...
SUBCOMMAND: # //cmd/kubelet:kubelet [action 'GoLink cmd/kubelet/linux_amd64_stripped/kubelet']
(cd /usr/local/google/home/bentheelder/.cache/bazel/_bazel_bentheelder/7be351c572778a98a0114cc199d97d3b/execroot/__main__ && \
  exec env - \
    CGO_ENABLED=1 \
    GOARCH=amd64 \
    GOOS=linux \
    GOROOT='bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~' \
    GOROOT_FINAL=GOROOT \
    PATH=/usr/bin \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/linux_amd64_stripped/link -go external/go_sdk/bin/go -dep '//cmd/kubelet/app:go_default_library=k8s.io/kubernetes/cmd/kubelet/app=bazel-out/k8-fastbuild/bin/cmd/kubelet/app/linux_amd64_stripped/go_default_library~/k8s.io/kubernetes/cmd/kubelet/app.a' [... many args elided ...] -stamp bazel-out/stable-status.txt -stamp bazel-out/volatile-status.txt -o bazel-out/k8-fastbuild/bin/cmd/kubelet/linux_amd64_stripped/kubelet -main 'bazel-out/k8-fastbuild/bin/cmd/kubelet/linux_amd64_stripped/kubelet~/k8s.io/kubernetes/cmd/kubelet.a' -- -extld /usr/bin/gcc -w -extldflags '-fuse-ld=gold -Wl,-no-as-needed -Wl,-z,relro,-z,now -B/usr/bin -B/usr/bin -pass-exit-codes -lstdc++ -lm')
Target //cmd/kubelet:kubelet up-to-date:
  bazel-bin/cmd/kubelet/linux_amd64_stripped/kubelet
```

which shows `-lstdc++ -lm` being added for some reason.
After some code sleuthing, I found this [suspicious line in the bazel tree](
https://github.com/bazelbuild/bazel/blob/5e5ee0db8f1925bcf9ac39232faa0622966d7e2a/tools/cpp/unix_cc_configure.bzl#L297):
```python
            get_env_var(repository_ctx, "BAZEL_LINKOPTS", "-lstdc++:-lm", False),
```

I tried setting `BAZEL_LINKOPTS` to an empty string, which didn't work:
```console
 $ BAZEL_LINKOPTS="" bazel build cmd/kubelet
INFO: Build options have changed, discarding analysis cache.
INFO: Analysed target //cmd/kubelet:kubelet (1 packages loaded).
INFO: Found 1 target...
ERROR: Process terminated by signal 15
Target //cmd/kubelet:kubelet failed to build
Unhandled exception thrown during build; message: Unrecoverable error while evaluating node 'ActionLookupData{actionLookupKey=//vendor/github.com/docker/docker/pkg/term:go_default_library._cgo_.o BuildConfigurationValue.Key[70f02cd64c445551b9cc0d353e7b00f5] false, actionIndex=2}' (requested by nodes 'vendor/github.com/docker/docker/pkg/term/go_default_library._cgo_.o-2.params //vendor/github.com/docker/docker/pkg/term:go_default_library._cgo_.o BuildConfigurationValue.Key[70f02cd64c445551b9cc0d353e7b00f5] false')
INFO: Elapsed time: 2.186s, Critical Path: 0.01s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData{actionLookupKey=//vendor/github.com/docker/docker/pkg/term:go_default_library._cgo_.o BuildConfigurationValue.Key[70f02cd64c445551b9cc0d353e7b00f5] false, actionIndex=2}' (requested by nodes 'vendor/github.com/docker/docker/pkg/term/go_default_library._cgo_.o-2.params //vendor/github.com/docker/docker/pkg/term:go_default_library._cgo_.o BuildConfigurationValue.Key[70f02cd64c445551b9cc0d353e7b00f5] false')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:477)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:355)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 0
        at java.base/java.lang.StringLatin1.charAt(Unknown Source)
        at java.base/java.lang.String.charAt(Unknown Source)
...
```

So then I tried `BAZEL_LINKOPTS="-lm"`, but that didn't work either:
```console
$ BAZEL_LINKOPTS="-lm" bazel build cmd/kubelet
...
  /usr/bin/gcc -o bazel-out/host/bin/external/com_google_protobuf/js_embed '-fuse-ld=gold' -Wl,-no-as-needed -Wl,-z,relro,-z,now -B/usr/bin -B/usr/bin -pass-exit-codes -Wl,--gc-sections -Wl,-S -Wl,@bazel-out/host/bin/external/com_google_protobuf/js_embed-2.params)

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/host/bin/external/com_google_protobuf/_objs/js_embed/embed.o:embed.cc:function CEscape(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&): error: undefined reference to 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const*, unsigned long)'
bazel-out/host/bin/external/com_google_protobuf/_objs/js_embed/embed.o:embed.cc:function CEscape(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&): error: undefined reference to 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const*, unsigned long)'
bazel-out/host/bin/external/com_google_protobuf/_objs/js_embed/embed.o:embed.cc:function CEscape(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&): error: undefined reference to 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const*, unsigned long)'
bazel-out/host/bin/external/com_google_protobuf/_objs/js_embed/embed.o:embed.cc:function CEscape(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&): error: undefined reference to 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const*, unsigned long)'
...
``` 

So it seems like we really do need `-lstdc++` for code which is built as C++, but Go should be treated as C. Is there any way to get rules_go / bazel to treat these differently?

--- a/go/private/actions/link.bzl
+++ b/go/private/actions/link.bzl
@@ -66,7 +66,11 @@ def emit_link(
     config_strip = len(go._ctx.configuration.bin_dir.path) + 1
     pkg_depth = executable.dirname[config_strip:].count("/") + 1
 
-    extldflags = list(go.cgo_tools.linker_options)
+    # Exclude -lstdc++ from link options. We don't want to link against it
+    # unless we actually have some C++ code. _cgo_codegen will include it
+    # in archives via CGO_LDFLAGS if it's needed.
+    extldflags = [f for f in go.cgo_tools.linker_options if f not in ("-lstdc++", "-lc++")]
+
     if go.coverage_enabled:
         extldflags.append("--coverage")
     gc_linkopts, extldflags = _extract_extldflags(gc_linkopts, extldflags)

--- a/go/private/rules/cgo.bzl
+++ b/go/private/rules/cgo.bzl
@@ -188,6 +188,12 @@ def _cgo_codegen_impl(ctx):
         objc_outs.append(gen_file)
         builder_args.add_all(["-src", gen_file.path + "=" + src.path])
 
+    # Filter out -lstdc++ in CGO_LDFLAGS if we don't have any C++ code. This
+    # also gets filtered out in link.bzl.
+    have_cc = len(source.cxx) + len(source.objc) + len(ctx.attr.deps) > 0
+    if not have_cc:
+        linkopts = [o for o in linkopts if o not in ("-lstdc++", "-lc++")]
+
     tool_args.add_all(["-objdir", out_dir])
 
     inputs = sets.union(ctx.files.srcs, go.crosstool, go.sdk.tools, go.stdlib.libs)

--- a/tests/core/cgo/BUILD.bazel
+++ b/tests/core/cgo/BUILD.bazel
@@ -1,4 +1,4 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")
 
 go_test(
     name = "opts_test",
@@ -8,16 +8,16 @@ go_test(
 
 go_library(
     name = "opts",
-    cgo = True,
-    copts = ["-DRULES_GO_C"],
-    cxxopts = ["-DRULES_GO_CXX"],
-    cppopts = ["-DRULES_GO_CPP"],
     srcs = [
-        "add.h",
         "add.c",
         "add.cpp",
+        "add.h",
         "adder.go",
     ],
+    cgo = True,
+    copts = ["-DRULES_GO_C"],
+    cppopts = ["-DRULES_GO_CPP"],
+    cxxopts = ["-DRULES_GO_CXX"],
     importpath = "github.com/bazelbuild/rules_go/tests/core/cxx",
 )
 
@@ -31,12 +31,12 @@ go_test(
 go_library(
     name = "dylib_client",
     srcs = ["dylib_client.go"],
-    cgo = True,
     cdeps = select({
         "@io_bazel_rules_go//go/platform:darwin": [":darwin_imported_dylib"],
         "//conditions:default": [":linux_imported_dylib"],
         # TODO(jayconrod): Support windows, skip others.
     }),
+    cgo = True,
     importpath = "github.com/bazelbuild/rules_go/tests/core/cgo/dylib",
 )
 
@@ -66,4 +66,60 @@ genrule(
     outs = ["libimported.so"],
     cmd = "$(CC) -shared $< -o $@",
     tags = ["manual"],
+)
+
+go_test(
+    name = "cc_libs_test",
+    srcs = [
+        "cc_libs_darwin_test.go",
+        "cc_libs_linux_test.go",
+    ],
+    data = [
+        ":c_srcs",
+        ":cc_deps",
+        ":cc_srcs",
+        ":pure",
+    ],
+    deps = ["//go/tools/bazel:go_default_library"],
+)
+
+go_binary(
+    name = "pure",
+    srcs = ["main.go"],
+    out = "pure_bin",
+    cgo = True,
+    pure = "on",
+)
+
+go_binary(
+    name = "c_srcs",
+    srcs = [
+        "foo.c",
+        "main.go",
+    ],
+    out = "c_srcs_bin",
+    cgo = True,
+)
+
+go_binary(
+    name = "cc_srcs",
+    srcs = [
+        "bar.cc",
+        "main.go",
+    ],
+    out = "cc_srcs_bin",
+    cgo = True,
+)
+
+go_binary(
+    name = "cc_deps",
+    srcs = ["main.go"],
+    out = "cc_deps_bin",
+    cdeps = [":bar_dep"],
+    cgo = True,
+)
+
+cc_library(
+    name = "bar_dep",
+    srcs = ["bar.cc"],
 )

--- a/tests/core/cgo/README.rst
+++ b/tests/core/cgo/README.rst
@@ -13,3 +13,15 @@ dylib_test
 Checks that Go binaries can link against dynamic C libraries. Some libraries
 (especially those provided with ``cc_import``) may only have dynamic versions,
 and we should be able to link against them and find them at run-time.
+
+cc_libs_test
+------------
+
+Checks that Go binaries that include cgo code may or may not link against
+libstdc++, depending on how they're linked. This tests several binaries:
+
+* ``pure_bin`` - built in ``"pure"`` mode, should not depend on libstdc++.
+* ``c_srcs`` - has no C++ code in sources, should not depend on libstdc++.
+* ``cc_srcs`` - has some C++ code in sources, should depend on libstdc++.
+* ``cc_deps`` - depends on a ``cc_library``, should depend on libstdc++
+  because we don't know what's in it.

--- a/tests/core/cgo/bar.cc
+++ b/tests/core/cgo/bar.cc
@@ -1,0 +1,5 @@
+#include <iostream>
+
+void bar() {
+  std::cout << "bar" << std::endl;
+}

--- a/tests/core/cgo/cc_libs_darwin_test.go
+++ b/tests/core/cgo/cc_libs_darwin_test.go
@@ -1,0 +1,91 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cc_libs_test
+
+import (
+	"debug/macho"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+)
+
+func TestBinaries(t *testing.T) {
+	for _, test := range []struct {
+		shortPath string
+		wantLibs  map[string]bool
+	}{
+		{
+			shortPath: "tests/core/cgo/pure_bin",
+			wantLibs:  map[string]bool{"libSystem": false, "libc++": false},
+		}, {
+			shortPath: "tests/core/cgo/c_srcs_bin",
+			wantLibs:  map[string]bool{"libSystem": true, "libc++": false},
+		}, {
+			shortPath: "tests/core/cgo/cc_srcs_bin",
+			wantLibs:  map[string]bool{"libSystem": true, "libc++": true},
+		}, {
+			shortPath: "tests/core/cgo/cc_deps_bin",
+			wantLibs:  map[string]bool{"libSystem": true, "libc++": true},
+		},
+	} {
+		t.Run(path.Base(test.shortPath), func(t *testing.T) {
+			libs, err := listLibs(test.shortPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+			haveLibs := make(map[string]bool)
+			for _, lib := range libs {
+				haveLibs[lib] = true
+			}
+			for haveLib := range haveLibs {
+				if wantLib, ok := test.wantLibs[haveLib]; ok && !wantLib {
+					t.Errorf("unexpected dependency on library %q", haveLib)
+				}
+			}
+			for wantLib, want := range test.wantLibs {
+				if want && !haveLibs[wantLib] {
+					t.Errorf("wanted dependency on library %q", wantLib)
+				}
+			}
+		})
+	}
+}
+
+func listLibs(shortPath string) ([]string, error) {
+	binPath, err := bazel.Runfile(shortPath)
+	if err != nil {
+		return nil, err
+	}
+	f, err := macho.Open(binPath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	libs, err := f.ImportedLibraries()
+	if err != nil {
+		return nil, err
+	}
+	for i := range libs {
+		if pos := strings.LastIndexByte(libs[i], '/'); pos >= 0 {
+			libs[i] = libs[i][pos+1:]
+		}
+		if pos := strings.IndexByte(libs[i], '.'); pos >= 0 {
+			libs[i] = libs[i][:pos]
+		}
+	}
+	return libs, nil
+}

--- a/tests/core/cgo/cc_libs_linux_test.go
+++ b/tests/core/cgo/cc_libs_linux_test.go
@@ -1,0 +1,91 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cc_libs_test
+
+import (
+	"debug/elf"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+)
+
+func TestBinaries(t *testing.T) {
+	for _, test := range []struct {
+		shortPath string
+		wantLibs  map[string]bool
+	}{
+		{
+			shortPath: "tests/core/cgo/pure_bin",
+			wantLibs:  map[string]bool{"libc": false, "libstdc++": false},
+		}, {
+			shortPath: "tests/core/cgo/c_srcs_bin",
+			wantLibs:  map[string]bool{"libc": true, "libstdc++": false},
+		}, {
+			shortPath: "tests/core/cgo/cc_srcs_bin",
+			wantLibs:  map[string]bool{"libc": true, "libstdc++": true},
+		}, {
+			shortPath: "tests/core/cgo/cc_deps_bin",
+			wantLibs:  map[string]bool{"libc": true, "libstdc++": true},
+		},
+	} {
+		t.Run(path.Base(test.shortPath), func(t *testing.T) {
+			libs, err := listLibs(test.shortPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+			haveLibs := make(map[string]bool)
+			for _, lib := range libs {
+				haveLibs[lib] = true
+			}
+			for haveLib := range haveLibs {
+				if wantLib, ok := test.wantLibs[haveLib]; ok && !wantLib {
+					t.Errorf("unexpected dependency on library %q", haveLib)
+				}
+			}
+			for wantLib, want := range test.wantLibs {
+				if want && !haveLibs[wantLib] {
+					t.Errorf("wanted dependency on library %q", wantLib)
+				}
+			}
+		})
+	}
+}
+
+func listLibs(shortPath string) ([]string, error) {
+	binPath, err := bazel.Runfile(shortPath)
+	if err != nil {
+		return nil, err
+	}
+	f, err := elf.Open(binPath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	libs, err := f.ImportedLibraries()
+	if err != nil {
+		return nil, err
+	}
+	for i := range libs {
+		if pos := strings.LastIndexByte(libs[i], '/'); pos >= 0 {
+			libs[i] = libs[i][pos+1:]
+		}
+		if pos := strings.IndexByte(libs[i], '.'); pos >= 0 {
+			libs[i] = libs[i][:pos]
+		}
+	}
+	return libs, nil
+}

--- a/tests/core/cgo/foo.c
+++ b/tests/core/cgo/foo.c
@@ -1,0 +1,7 @@
+#include <stdio.h>
+
+void foo() {
+  printf("foo\n");
+}
+
+ 

--- a/tests/core/cgo/main.go
+++ b/tests/core/cgo/main.go
@@ -1,0 +1,3 @@
+package main
+
+func main() {}

